File: DP/javascript/2579.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const n = Number(input[0]);
dp = Array.from(Array(n + 1), () => new Array(2));
dp[1][0] = Number(input[1]);
dp[1][1] = 0;
if (n >= 2) {
  dp[2][0] = Number(input[2]);
  dp[2][1] = dp[1][0] + Number(input[2]);
}

const search = (n) => {
  if (dp[n][0] !== undefined) return dp[n];
  dp[n][0] = Math.max(...search(n - 2)) + Number(input[n]);
  dp[n][1] = Math.max(search(n - 1)[0]) + Number(input[n]);
  return dp[n];
};
search(n);
console.log(Math.max(...dp[n]));

File: DP/python/2579.py
n = int(input())
points = [int(input()) for _ in range(n)]
dp = [[0,0] for _ in range(n+1)]

dp[0][0] = points[0]
dp[0][1] = 0
if n>=2:
  dp[1][0] = points[1]
  dp[1][1] = dp[0][0] + points[1]

def search(i):
  if dp[i][0] : return dp[i]
  dp[i][0] = max(search(i-2)) + points[i]
  dp[i][1] = search(i-1)[0] + points[i]
  return dp[i]
print(max(search(n-1)))
